File: leetcode/100-999/160. Intersection of Two Linked Lists/solution_2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution 
{
    public:
        ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) 
        {
            set<ListNode*> nodes;
            ListNode* node;
            
            node = headA;
            while (node)
            {
                nodes.insert(node);
                node = node->next;
            }

            node = headB;
            while (node)
            {
                if (nodes.find(node) != nodes.end())
                    return node;

                node = node->next;
            }
            
            return nullptr;
        }
};

File: leetcode/100-999/160. Intersection of Two Linked Lists/solution_2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */

public class Solution
{
    public ListNode getIntersectionNode(ListNode headA, ListNode headB)
    {
        HashSet<ListNode> nodes = new HashSet<>();
        ListNode node = null;

        node = headA;
        while (node != null)
        {
            nodes.add(node);
            node = node.next;
        }
        
        node = headB;
        while (node != null)
        {
            if (nodes.contains(node))
                return node;
            
            node = node.next;
        }
        
        return null;
    }
}